Add an option for turning on/off compaction from apiserver in etcd3 mode

mitake opened the pull request:
…erver

**What this PR does / why we need it**:

This commit adds an option for controlling request of compaction to
etcd3 from apiserver. There is a situation that apiserver cannot fully
own its etcd cluster (e.g. sharing it with canal). In such a case,
apiserver should have limited access in terms of etcd's auth
functionality so it don't have a privilege to issue compaction
requests. It means that the compaction requests should be issued by
other component and apiserver's compaction requests are needless.

For such use cases, this commit adds a new flag
storagebackend.Config.DoCompaction. If the flag is true (default),
apiserver issues the compaction requests like current behaviour. If it
is false, apiserver doesn't issue the requests.

**Related issue (etcd)**
https://github.com/coreos/etcd/issues/8458
/cc @xiang90 @struz

**Release note:**
```release-note
Add --etcd-compaction-interval to apiserver for controlling request of compaction to etcd3 from apiserver.
```

mitake committed:
storage, etcd3: add an option for configuring interval of compaction requests from apiserver

This commit adds an option for controlling request of compaction to
etcd3 from apiserver. There is a situation that apiserver cannot fully
own its etcd cluster (e.g. sharing it with canal). In such a case,
apiserver should have limited access in terms of etcd's auth
functionality so it don't have a priviledge to issue compaction
requests. It means that the compaction requests should be issued by
other component and apiserver's compaction requests are needless.

For such use cases, this commit adds a new flag
`storagebackend.Config.CompactionInterval`. If the flag is non 0,
apiserver issues the compaction requests like current behaviour (the
default is 5 minutes). If it is 0, apiserver doesn't issue the
requests. It can be configured with a newly added option of apiserver
`--etcd-compaction-interval`.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -50,6 +50,9 @@ type Config struct {
 	Transformer value.Transformer
+
+	// DoCompaction is a flag to turn on or off requesting compaction from apiserver.
+	DoCompaction bool
```
I'm fine with exposing this as an option - I've already heard about customers who would like to control it themselves too.

I'm not sure if DoCompaction is the best name though.
I think that "auto-compaction" should be default and you should be able to opt-out. That would mean I would suggest changing to:
```
DisableCompaction
```
[with default false].

mitake replied:
Thanks for your comment. I'll rename the flag to `DisableCompaction`.

mitake replied:
@wojtek-t BTW can I add `DisableCompaction` to `EtcdOptions` instead of `storagebackend.Config`? It seems that we don't have a way of configuring `storagebackend.Config` by command line option or other ways.

wojtek-t replied:
Yeah - that is fine

mitake replied:
@wojtek-t sorry I found a very simple way to do this... please ignore the above comment

mitake commented:
@wojtek-t I renamed the field name to `DisableCompaction` and added an option of apiserver (`--etcd-disable-compaction`) for controlling the field. I'll test it on a real cluster tomorrow. Should I add some tests for the new option and field?

xiang90 commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -127,6 +127,10 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EncryptionProviderConfigFilepath, "experimental-encryption-provider-config", s.EncryptionProviderConfigFilepath,
 		"The file containing configuration for encryption providers to be used for storing secrets in etcd")
+
+	fs.BoolVar(&s.StorageConfig.DisableCompaction, "etcd-disable-compaction", s.StorageConfig.DisableCompaction,
```
probably we should make it clear that if this option is enabled, user MUST compact the underlying etcd3 storage themselves.

mitake replied:
It makes sense. I'll update the description later. Thanks.

mitake commented:
@xiang90 @wojtek-t I updated the description and briefly tested it and at least it wouldn't be breaking existing things. Could you trigger the tests? Or should I add new tests or a release note (sorry I'm not familiar with PR merging process of k8s...)?

smarterclayton commented:
So given that we may want to tune this for apichunking, I'd prefer to make
this field an interval with a default value and the value 0 mean
"compaction off".

On Fri, Sep 8, 2017 at 9:55 AM, Hitoshi Mitake <notifications@github.com>
wrote:

> @xiang90 <https://github.com/xiang90> @wojtek-t
> <https://github.com/wojtek-t> I updated the description and briefly
> tested it and at least it wouldn't be breaking existing things. Could you
> trigger the tests? Or should I add new tests or a release note (sorry I'm
> not familiar with PR merging process of k8s...)?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/51765#issuecomment-328109518>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p1HizaM9R1v1yoyoDk7qjXvYbzH2ks5sgUdDgaJpZM4PJuBH>
> .
>

mitake commented:
@smarterclayton what does "apichunking" mean (sorry I'm not familiar with k8s terminologies)?
Do you mean that the option should be able to specify the interval of compaction (`etcd3.compactInterval`)? If this understanding is correct, I agree with your opinion so will change the flag from bool to duration.

wojtek-t commented:
In this context apichunking means "pagination of the LIST responses".
I agree with Clayton`s suggestion to make it a duration.

xiang90 commented:
@mitake 

The API server specifies etcd revision to get a consistent snapshot for pagination. It can achieve isolation with it.

However, if compaction compacts a revision that a client is ranging over, the client has to abort the range and start over.

Controlling the compaction interval can hint the client how fast it should finish the range without getting a abort.

mitake commented:
@wojtek-t @xiang90 I see, thanks a lot for your explanation! I'll update this PR later.

mitake commented:
@xiang90 @wojtek-t @smarterclayton I changed the type of option from bool to duration and the name to `--etd-compaction-interval`. In my brief test, the option seems to be working well. Could you take a look?

smarterclayton commented:
Thank you for changing - high level this looks ok to me.

wojtek-t commented:
/lgtm

/ok-to-test

mitake commented:
@wojtek-t I updated this PR for resolving the unit test failure (`TestAddFlags`). Other failures related to bazel couldn't be reproduced on my local env. Could you take a look again?

mitake commented:
@k8s-ci-robot /retest

mitake commented:
@k8s-ci-robot /retest

mitake commented:
It seems that `pull-kubernetes-kubemark-e2e-gce` is failing because of disk full. I'm not sure this should be categorized as fleaky test. Should I open an issue for this?

wojtek-t commented:
@mitake - the problem with full disk is not job specific - it's about test infrastructure. We are seeing it in many different jobs on many different PRs.

wojtek-t commented:
/lgtm
/approve no-issue

mitake commented:
@wojtek-t ah sorry I added a release note to the PR description. Should I remove it (you added release-note-none tag)?

mitake commented:
@wojtek-t and I understand the situation about the disk full issue, thanks.

wojtek-t commented:
Actually, I think this release note makes sense - changed the label.

mitake commented:
I see, thanks for updating the label.

mitake commented:
/retest

mitake commented:
/retest (the unit test isn't triggered yet)

BTW, can this be pushed to the submit queue because v1.8 is now released?

wojtek-t commented:
/retest

wojtek-t commented:
/test pull-kubernetes-unit
